vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_context.h

```diff
@@ -0,0 +1,272 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_BAT_LEDGER_CONTEXT_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_BAT_LEDGER_CONTEXT_H_
+
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#include "base/bind.h"
+#include "base/location.h"
+#include "base/memory/weak_ptr.h"
+#include "bat/ledger/ledger.h"
+#include "bat/ledger/public/interfaces/ledger.mojom.h"
+
+namespace ledger {
+
+class LedgerImpl;
+class LedgerClient;
+
+// Represents a per-user running instance of the BAT client engine. It serves as
```

Comment:
not sure I'm following, LedgerImpl is already per-user (profile) and how can singletons be per-user?

Reply:
`LedgerImpl` is exactly the problem that this abstraction is attempting to solve: it is a manual dependency hub in which all of its subcomponents store a raw pointer back to `LedgerImpl`. It is non-scalable, non-compositional, non-testable, and, have you seen all of the raw `this` pointers?

Re "singleton": perhaps a misuse, but the intended meaning here is one instance of the component per context.